naga/src/front/atomic_upgrade.rs

```diff
@@ -0,0 +1,338 @@
+            .module
+            .types
+            .get_handle(type_handle)
+            .map_err(Error::MissingHandle)?
+            .clone();
+        padding.debug("type: ", &type_);
+
+        let new_inner = match type_.inner {
+            TypeInner::Scalar(scalar) => {
+                log::trace!("{padding}hit the scalar leaf, replacing with an atomic");
+                TypeInner::Atomic(scalar)
+            }
+            TypeInner::Pointer { base, space } => TypeInner::Pointer {
+                base: self.upgrade_type(base)?,
+                space,
+            },
+            TypeInner::Array { base, size, stride } => TypeInner::Array {
+                base: self.upgrade_type(base)?,
+                size,
+                stride,
+            },
+            TypeInner::Struct { members, span } => TypeInner::Struct {
+                members: {
+                    let mut new_members = vec![];
+                    for member in members.iter().cloned() {
+                        let StructMember {
+                            name,
+                            ty,
+                            binding,
+                            offset,
+                        } = member;
+                        new_members.push(StructMember {
+                            name,
+                            ty: self.upgrade_type(ty)?,
+                            binding,
+                            offset,
+                        });
+                    }
+                    new_members
+                },
+                span,
+            },
+            TypeInner::BindingArray { base, size } => TypeInner::BindingArray {
+                base: self.upgrade_type(base)?,
+                size,
+            },
+            n => n,
+        };
+
+        let new_type_handle = self.module.types.insert(
+            Type {
+                name: type_.name,
+                inner: new_inner,
+            },
+            self.module.types.get_span(type_handle),
+        );
+        padding.debug("new_type: ", new_type_handle);
+        Ok(new_type_handle)
+    }
+
+    fn upgrade_global_variable(
+        &mut self,
+        handle: Handle<GlobalVariable>,
+    ) -> Result<Handle<GlobalVariable>, Error> {
+        let padding = self.inc_padding();
+        padding.debug("upgrading global variable: ", handle);
+
+        let var = self.module.global_variables.try_get(handle)?.clone();
+        padding.debug("global variable:", &var);
+
+        let new_var = GlobalVariable {
+            name: var.name.clone(),
+            space: var.space,
+            binding: var.binding.clone(),
+            ty: self.upgrade_type(var.ty)?,
+            init: self.upgrade_opt_expression(None, var.init)?,
+        };
+        if new_var != var {
+            padding.debug("new global variable: ", &new_var);
+            let span = self.module.global_variables.get_span(handle);
+            let new_handle = self.module.global_variables.append(new_var, span);
```

Comment:
You can't duplicate the global variable here - you need to modify the type of the existing `GlobalVariable` in place. This actually makes things simpler, because you don't need to change any `Expression`s, they're all pointing at the right `GlobalVariable` already.

Reply:
Oh, wow. That one little change (modifying the global var in place) actually makes the generated module validate!